core/trino-main/src/main/java/io/trino/operator/join/unspilled/JoinProbe.java

```diff
@@ -108,23 +110,79 @@ public Page getPage()
         return page;
     }
 
-    private boolean currentRowContainsNull()
+    private long[] fillCache()
     {
-        for (int i = 0; i < probePage.getChannelCount(); i++) {
-            if (probePage.getBlock(i).isNull(position)) {
-                return true;
+        long[] joinPositionCache = new long[positionCount];
+        Arrays.fill(joinPositionCache, -1);
+        if (!nullableBlocks.isEmpty()) {
+            boolean[] isNull = new boolean[positionCount];
+            int nonNullCount = getIsNull(isNull);
+            if (nonNullCount < positionCount) {
+                // We only store positions that are not null
+                int[] positions = new int[nonNullCount];
+                nonNullCount = 0;
+                for (int i = 0; i < positionCount; i++) {
+                    if (!isNull[i]) {
+                        positions[nonNullCount] = i;
+                    }
+                    // This way less code is in the if branch and CPU should be able to optimize branch prediction better
+                    nonNullCount += isNull[i] ? 0 : 1;
+                }
+                long[] packedPositionCache;
+                if (probeHashBlock != null) {
+                    long[] hashes = new long[nonNullCount];
+                    for (int i = 0; i < nonNullCount; i++) {
+                        hashes[i] = BIGINT.getLong(probeHashBlock, positions[i]);
+                    }
+                    packedPositionCache = lookupSource.getJoinPosition(positions, probePage, page, hashes);
+                }
+                else {
+                    packedPositionCache = lookupSource.getJoinPosition(positions, probePage, page);
+                }
+                // Unpack
+                nonNullCount = 0;
+                for (int i = 0; i < positionCount; i++) {
+                    if (!isNull[i]) {
+                        joinPositionCache[i] = packedPositionCache[nonNullCount];
+                    }
+                    // This way less code is in the if branch and CPU should be able to optimize branch prediction better
+                    nonNullCount += isNull[i] ? 0 : 1;
+                }
+                return joinPositionCache;
+            } // else fall back to non-null path
+        }
+        int[] positions = new int[positionCount];
+        for (int i = 0; i < positionCount; i++) {
+            positions[i] = i;
+        }
+        if (probeHashBlock != null) {
+            long[] hashes = new long[positionCount];
+            for (int i = 0; i < positionCount; i++) {
+                hashes[i] = BIGINT.getLong(probeHashBlock, i);
             }
+            return lookupSource.getJoinPosition(positions, probePage, page, hashes);
+        }
+        else {
+            return lookupSource.getJoinPosition(positions, probePage, page);
         }
-        return false;
     }
 
-    private static boolean probeMayHaveNull(Page probePage)
+    private int getIsNull(boolean[] isNull)
```

Comment:
nit: when `results` is `out` parameter, this could look like https://github.com/starburstdata/trino/commit/9bbe9bd2675c22a498c121315fb589be3c73096c